File: sdks/typescript/src/types/channel.ts
/**
 * Channel types — modeled on the backend wire contract (snake_case).
 * Credentials are write-only and never present on a response.
 */

import type { ChannelType, PaginationParams } from './common';

/** Live connection state (wire field `connection_status`), distinct from `enabled`. */
export type ConnectionStatus = 'disconnected' | 'connecting' | 'connected' | 'error';

/** WhatsApp Business App + Cloud API coexistence state. */
export type CoexistenceStatus =
  | 'inactive'
  | 'pending'
  | 'active'
  | 'warning'
  | 'disconnected';

/**
 * Channel `config` is a flat string map on the wire (e.g.
 * `{ phone_number_id: "...", waba_id: "..." }`). Secret values are redacted to
 * `"__redacted__"` in responses.
 */
export type ChannelConfig = Record<string, string>;

export interface Channel {
  id: string;
  tenant_id: string;
  type: ChannelType;
  name: string;
  identifier?: string;
  /** System-level enable flag (distinct from connection_status). */
  enabled: boolean;
  connection_status: ConnectionStatus;
  config?: ChannelConfig;
  webhook_url?: string;
  created_at: string;
  updated_at: string;
  // WhatsApp coexistence
  is_coexistence?: boolean;
  waba_id?: string;
  last_echo_at?: string;
  coexistence_status?: CoexistenceStatus;
  message_template_namespace?: string;
}

// Request types

/**
 * `config` holds non-secret settings (phone_number_id, waba_id, ...).
 * `credentials` holds secrets (access_token, bot_token, ...) — stored encrypted,
 * never returned. `webhook_url` is the external endpoint Linktor delivers signed
 * inbound/status events to.
 */
export interface CreateChannelInput {
  name: string;
  type: ChannelType;
  identifier?: string;
  config?: Record<string, string>;
  credentials?: Record<string, string>;
  webhook_url?: string;
}

/**
 * Update reuses the create body shape. `credentials`, when present, replace the
 * stored secrets; omit it (or send the redacted placeholder) to keep them.
 */
export interface UpdateChannelInput {
  name?: string;
  identifier?: string;
  config?: Record<string, string>;
  credentials?: Record<string, string>;
  webhook_url?: string;
}

/**
 * Result of connecting a channel. For WhatsApp Web-style linking, `qr_code`
 * carries the payload to render and `expires_in` its lifetime in seconds — call
 * `connect` again to refresh an expired code. `pair_code` is the phone-linking
 * code. When `passkey_required` is true the account is passkey-locked and must
 * be linked by signing `passkey_challenge` (submit via the passkey endpoint),
 * not by QR.
 */
export interface ConnectResult {
  /** Present when the response carries a channel; the wire may send null. */
  channel?: Channel;
  qr_code?: string;
  expires_in?: number;
  pair_code?: string;
  passkey_required?: boolean;
  passkey_challenge?: unknown;
}

/** Body for requesting a WhatsApp pairing code. */
export interface PairCodeInput {
  phone_number: string;
}

export interface ListChannelsParams extends PaginationParams {
  type?: ChannelType;
  status?: ConnectionStatus;
  search?: string;
}

export interface ChannelCapabilities {
  supportedContentTypes: string[];
  supportsMedia: boolean;
  supportsButtons: boolean;
  supportsLists: boolean;
  supportsTemplates: boolean;
  supportsLocation: boolean;
  maxMessageLength: number;
  maxMediaSize: number;
}

File: sdks/go/types/channel.go
package types

import (
	"encoding/json"
	"time"
)

// ConnectionStatus is the live connection state of a channel (wire field
// "connection_status"). It is distinct from Enabled, which is the system-level
// enable flag.
type ConnectionStatus string

const (
	ConnectionStatusDisconnected ConnectionStatus = "disconnected"
	ConnectionStatusConnecting   ConnectionStatus = "connecting"
	ConnectionStatusConnected    ConnectionStatus = "connected"
	ConnectionStatusError        ConnectionStatus = "error"
)

// CoexistenceStatus is the WhatsApp Business App + Cloud API coexistence state.
type CoexistenceStatus string

const (
	CoexistenceStatusInactive     CoexistenceStatus = "inactive"
	CoexistenceStatusPending      CoexistenceStatus = "pending"
	CoexistenceStatusActive       CoexistenceStatus = "active"
	CoexistenceStatusWarning      CoexistenceStatus = "warning"
	CoexistenceStatusDisconnected CoexistenceStatus = "disconnected"
)

// Channel mirrors the backend wire shape exactly (snake_case). Credentials are
// write-only and never present on a response.
type Channel struct {
	ID                       string            `json:"id"`
	TenantID                 string            `json:"tenant_id"`
	Type                     ChannelType       `json:"type"`
	Name                     string            `json:"name"`
	Identifier               string            `json:"identifier,omitempty"`
	Enabled                  bool              `json:"enabled"`
	ConnectionStatus         ConnectionStatus  `json:"connection_status"`
	Config                   map[string]string `json:"config,omitempty"`
	WebhookURL               string            `json:"webhook_url,omitempty"`
	CreatedAt                time.Time         `json:"created_at"`
	UpdatedAt                time.Time         `json:"updated_at"`
	IsCoexistence            bool              `json:"is_coexistence,omitempty"`
	WABAID                   string            `json:"waba_id,omitempty"`
	LastEchoAt               *time.Time        `json:"last_echo_at,omitempty"`
	CoexistenceStatus        CoexistenceStatus `json:"coexistence_status,omitempty"`
	MessageTemplateNamespace string            `json:"message_template_namespace,omitempty"`
}

// CreateChannelInput for creating channels.
//
// Config holds non-secret settings (e.g. phone_number_id, waba_id). Credentials
// holds secrets (e.g. access_token, bot_token) — the server stores them
// encrypted and never returns them. WebhookURL is the external endpoint Linktor
// delivers signed inbound/status events to for this channel.
type CreateChannelInput struct {
	Name        string            `json:"name"`
	Type        ChannelType       `json:"type"`
	Identifier  string            `json:"identifier,omitempty"`
	Config      map[string]string `json:"config,omitempty"`
	Credentials map[string]string `json:"credentials,omitempty"`
	WebhookURL  string            `json:"webhook_url,omitempty"`
}

// UpdateChannelInput for updating channels. Credentials, when present, replace
// the stored secrets; omit it (or send the redacted placeholder) to leave them
// untouched. Update reuses the create body shape server-side.
type UpdateChannelInput struct {
	Name        string            `json:"name,omitempty"`
	Identifier  string            `json:"identifier,omitempty"`
	Config      map[string]string `json:"config,omitempty"`
	Credentials map[string]string `json:"credentials,omitempty"`
	WebhookURL  string            `json:"webhook_url,omitempty"`
}

// ConnectResult is the result of connecting a channel. For WhatsApp Web-style
// linking it carries the QR payload (QRCode) to render and its lifetime
// (ExpiresIn seconds); poll Connect again to refresh an expired code. PairCode
// is the phone-linking code when pairing by number. PasskeyRequired is true when
// the account is passkey-locked and must be linked by signing PasskeyChallenge
// (submit the assertion via SubmitPasskeyResponse), not by QR.
type ConnectResult struct {
	Channel          *Channel        `json:"channel"`
	QRCode           string          `json:"qr_code,omitempty"`
	ExpiresIn        int             `json:"expires_in,omitempty"`
	PairCode         string          `json:"pair_code,omitempty"`
	PasskeyRequired  bool            `json:"passkey_required,omitempty"`
	PasskeyChallenge json.RawMessage `json:"passkey_challenge,omitempty"`
}

// PairCodeInput requests a WhatsApp pairing code for a phone number.
type PairCodeInput struct {
	PhoneNumber string `json:"phone_number"`
}

// ListChannelsParams for list request. Status/Type/Search are query filters.
type ListChannelsParams struct {
	PaginationParams
	Type   ChannelType      `json:"type,omitempty"`
	Status ConnectionStatus `json:"status,omitempty"`
	Search string           `json:"search,omitempty"`
}
